content/docs/testing-recipes.md

```diff
@@ -205,17 +205,30 @@ it("renders user data", async () => {
     address: "123, Charming Avenue"
   };
 
-  jest.spyOn(global, "fetch").mockImplementation(() =>
-    Promise.resolve({
-      json: () => Promise.resolve(fakeUser)
-    })
+  // Setup a mocked version of fetch that can be manually resolved:
+  let resolve;
+  jest.spyOn(global, "fetch").mockImplementation(
```

Comment:
Hoe about using jest.mockResolvedValue here or jest.mockResolvedValueOnce inline?

Reply:
[Here's what I recommend](https://github.com/kentcdodds/bookshelf/blob/4057cf5d5d6c0c3d40868abb804ffe3acf850c97/src/setupTests.js#L3-L13) (and I recommend doing this in a global setup, but it's fine to just have it in the test file for this example like this I think):

```javascript
beforeEach(() => {
  jest.spyOn(window, 'fetch').mockImplementation((...args) => {
    console.warn('window.fetch is not mocked for this call', ...args)
    return Promise.reject(new Error('This must be mocked!'))
  })
})

afterEach(() => {
  window.fetch.mockRestore()
})
```

Then in your test, you can just rely on the fact that `window.fetch` is a mock function. As for making it work nicely with `act` and resolving the promise before continuing things, your current implementation is fine. What would you think of doing something like making a little deferred function:

```javascript
function deferred() {
  let resolve, reject
  const promise = new Promise((res, rej) => {
    resolve = res
    reject = rej
  })
  return {resolve, reject, promise}
}
```

Then you could use that in the test like so:

```javascript
test('renders user data', async () => {
  // ...
  const {resolve, promise} = deferred()
  window.fetch.mockImplementationOnce(() => {
    return Promise.resolve({json: () => promise})
  })
  // ...

  resolve(fakeUser)
  await act(promise)
  // ...
```

I _think_ that should work.

This is so much easier with RTL, because you have the `find*` queries and `wait` utilities.

```js
// no need for a deferred utility or mutable variable references

test('renders user data', async () => {
  // ...
  window.fetch.mockImplementationOnce(() => {
    return Promise.resolve({json: () => Promise.resolve(fakeUser)})
  })
  // ...

  await wait(() => expect(container.querySelector("summary").textContent).toBe(fakeUser.name));
  // ...
```

Or even better with [the changes I was recommending for the RTL version](https://github.com/threepointone/reactjs.org/pull/1/files):

```javascript
test("renders user data", async () => {
  const fakeUser = {
    name: "Joni Baez",
    age: "32",
    address: "123, Charming Avenue",
  };
  window.fetch.mockImplementationOnce(() =>
    Promise.resolve({
      json: () => Promise.resolve(fakeUser),
    }),
  );

  const { getByLabelText, getByText } = render(<User id="123" />);

  expect(window.fetch).toHaveBeenCalledTimes(1);
  expect(window.fetch).toHaveBeenCalledWith("/123");

  // in any real world UI (pre-suspense as we are now), you'd have a loading indicator like so:
  await waitForElementToBeRemoved(() => getByText(/loading/i));

  // this requires aria-label to be added to the elements which is good for a11y
  expect(getByLabelText(/name/i)).toHaveTextContent(fakeUser.name);
  expect(getByLabelText(/age/i)).toHaveTextContent(fakeUser.age);
  expect(getByLabelText(/address/i)).toHaveTextContent(fakeUser.address);
});
```

**In any case**. If you want to change as little as possible, then I suggest at least adding this:

```javascript
beforeEach(() => {
  jest.spyOn(window, 'fetch').mockImplementation((...args) => {
    console.warn('window.fetch is not mocked for this call', ...args)
    return Promise.reject(new Error('This must be mocked!'))
  })
})

afterEach(() => {
  window.fetch.mockRestore()
})
```

Reason being, if this test fails, then the `global.fetch.mockRestore();` (which is currently in this test) would not run and the other tests in this file will fail in odd ways.

Sorry for butting in with opinions 😅 Feel free to ignore.

Reply:
all good feedback. I'll think some more.